soc: realrek: ec: Implement power saving

JasonLin-RealTek opened the pull request:
Impelment RTS5912 power saving

JasonLin-RealTek commented on soc/realtek/ec/rts5912/power.c:
```diff
@@ -50,9 +49,41 @@ static void rts5912_heavy_sleep(void)
+
+		pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+		k_work_reschedule(&rx_refresh_timeout_work, delay);
+	}
```
Thanks, we use pm_state_notify now.

fabiobaltieri commented on dts/arm/realtek/ec/rts5912.dtsi:
```diff
@@ -276,7 +276,7 @@
 			reg = <0x40010100 0x100>;
 			reg-shift = <2>;
-			clock-frequency = <25000000>;
+			clock-frequency = <100000000>;
```
You can use `DT_FREQ_M(100)` here, bit easier to read

JasonLin-RealTek replied:
Fixed

fabiobaltieri commented on dts/bindings/pinctrl/realtek,rts5912-pinctrl.yaml:
```diff
@@ -51,3 +54,13 @@ child-binding:
       description: |
         "fast" — Fast Frequency Slew Rate
         "slow" — Slow Frequency Slew Rate
+    interrupt-trigger:
+      type: string
+      description: |
+        Interrupt trigger mode.
+      enum:
+        - positive-edge_trigger
+        - negative-edge_trigger
+        - dual-edge_trigger
+        - low-level_trigger
+        - high-level_trigger
```
is this actually used anywhere? if it's going to be used on a followup PR then it should probably be added there together with whatever code ends up using it

JasonLin-RealTek replied:
We move it to the commit for serial which will use the gpio to be the wakeup reference.

JasonLin-RealTek replied:
We will remove every pinctrl modify on latest version, but unfortunately the timing issue for uart-gpio wake up is very critical. Therefore, we need to add the vendor function in the gpio driver to support this situation.

fabiobaltieri commented on drivers/serial/uart_realtek_rts5912.c:
```diff
@@ -31,6 +34,73 @@ struct uart_rts5912_device_config {
 };
 
+#if defined(CONFIG_PM)
+#include <reg/reg_gpio.h>
+#define RTS5912_RX_WAKEUP (DT_PROP(DT_INST(0, realtek_rts5912_uart), rx_wakeup_pin))
+#define RTS5912_UART_RX_GPIO                                                                       \
+	((volatile uint32_t *)(DT_REG_ADDR(DT_NODELABEL(gpioa)) + RTS5912_RX_WAKEUP * 4))
```
wondering if it couldn't be described as a `gpios` property rather than the raw number, like `rx-wakeup-gpios = <&gpioh 1 0>`

JasonLin-RealTek replied:
We remove the gpio detect here, and use uart rx to delay the sleep.

JasonLin-RealTek replied:
We use the gpios now, thanks!

JasonLin-RealTek committed:
soc: realtek: ec: Implement power saving

Impelment RTS5912 power saving (heavy sleep)
Remove the power state of "suspend_to_ram"

Signed-off-by: Lin Yu-Cheng <lin_yu_cheng@realtek.com>

JasonLin-RealTek committed:
serial: rts5912: implement power management

support uart wake up function

Signed-off-by: Lin Yu-Cheng <lin_yu_cheng@realtek.com>

JasonLin-RealTek commented on soc/realtek/ec/rts5912/soc.c:
```diff
@@ -28,15 +30,22 @@ bool z_arm_on_enter_cpu_idle(void)
+	if (IS_ENABLED(CONFIG_PM)) {
+		/* turn off TRNG/ OTP/ LDO3 power */
+		*(volatile uint32_t*)(0x40010800UL + 0x600UL + 0x20) = 0ul;
+		*(volatile uint32_t*)(0x40010800UL + 0x680UL + 0x24) = 0ul;
```
Thanks, we add the description define in the top of this code

JasonLin-RealTek commented on drivers/serial/uart_realtek_rts5912.c:
```diff
@@ -50,6 +120,12 @@ static int rts5912_uart_init(const struct device *dev)
+#if defined(CONFIG_PM)
+	k_work_init_delayable(&rx_refresh_timeout_work, uart_rts5912_rx_refresh_timeout);
+	pm_notifier_register(&notifier);
+	uart_irq_callback_set(DEVICE_DT_GET(DT_NODELABER(uart0)), uart_rx_wait);
```
We only want to check the zephyr console, therefore we change from uart0 to zephyr_console

JasonLin-RealTek commented on drivers/serial/uart_realtek_rts5912.c:
```diff
@@ -31,6 +34,73 @@ struct uart_rts5912_device_config {
+	.state_exit = notify_pm_state_exit,
+}
+
+void uart_rx_wait(const struct device *dev, void *user_data)
```
fixed

dsseng commented:
`soc: realrek: ec: Implement power saving` <- `realrek` is likely a typo, please fixup

It would be nicer if the `driver: serial: implement uart PM function` title is sth like `serial: rts5912: implement power management`, this would be more informative. That's just a suggestion from me as another contributor, so use your judgement and formulate commit as you feel

fabiobaltieri commented on soc/realtek/ec/common/pinctrl_soc.h:
```diff
@@ -50,9 +52,7 @@ typedef uint32_t pinctrl_soc_pin_t;
-	{                                                                                          \
-		DT_FOREACH_PROP_ELEM(node_id, prop, Z_PINCTRL_STATE_PIN_INIT)                      \
-	}
+	{DT_FOREACH_PROP_ELEM(node_id, prop, Z_PINCTRL_STATE_PIN_INIT)}
```
arguabely it's not, but we've been fairly loose on macros (unfortunately)

JasonLin-RealTek replied:
Fixed

JasonLin-RealTek committed:
gpio: rts5912: implement power management

support gpio wake up function.
change init stage from POST_KERNEL to PRE_KERNEL_1
because uart wrap init need to use the gpio functions.

Signed-off-by: Lin Yu-Cheng <lin_yu_cheng@realtek.com>

JasonLin-RealTek commented on drivers/serial/uart_realtek_rts5912.c:
```diff
@@ -31,6 +34,60 @@ struct uart_rts5912_device_config {
+{
+	ARG_UNUSED(user_data);
+	if (uart_irq_rx_ready(dev)) {
+		k_timeout_t delay = K_MSEC(15000);
```
We use CONFIG_UART_CONSOLE_INPUT_EXPIRED_TIMEOUT to represent the time.

JasonLin-RealTek commented on soc/realtek/ec/rts5912/soc.c:
```diff
@@ -28,15 +36,23 @@ bool z_arm_on_enter_cpu_idle(void)
+		*(volatile uint32_t *)(RTS5912_PUF_ADDRESS + RTS5912_TRNG_OFFSET +
+				       RTS5912_RNG_RUN) = 0ul;
+		*(volatile uint32_t *)(RTS5912_PUF_ADDRESS + RTS5912_OTP_OFFSET +
+				       RTS5912_CFG_PDSTB) = 0ul;
```
Thanks for the advise, we use sys_write32 here now.

JasonLin-RealTek commented on drivers/clock_control/clock_control_rts5912_sccon.c:
```diff
@@ -53,8 +53,7 @@ static int rts5912_periph_clock_control(const struct device *dev, clock_control_
 			sys_reg->UARTCLK = BIT(clk_idx);
-		} else {
-			sys_reg->UARTCLK &= ~BIT(clk_idx);
+			sys_reg->UARTCLK |= SYSTEM_UARTCLK_SRC_Msk;
```
Typo, Fixed.

JasonLin-RealTek commented on soc/realtek/ec/rts5912/soc.c:
```diff
@@ -8,11 +8,19 @@
+#define RTS5912_TRNG_OFFSET 0x600UL
+#define RTS5912_OTP_OFFSET  0x680UL
+#define RTS5912_CFG_PDSTB   0x0024UL
+#define RTS5912_RNG_RUN     0x0020UL
```
We have to` reg/reg_system` ,thanks.

JasonLin-RealTek commented on dts/bindings/pinctrl/realtek,rts5912-pinctrl.yaml:
```diff
@@ -35,6 +35,9 @@ child-binding:
       type: int
       required: true
       description: Pinmux selection
+    interrupt-enable:
```
We removed it, thanks.

fabiobaltieri commented on drivers/gpio/gpio_rts5912.c:
```diff
@@ -564,7 +593,7 @@ static DEVICE_API(gpio, gpio_rts5912_driver_api) = {
 		.num_pins = DT_INST_PROP(id, ngpios),                                              \
 	};                                                                                         \
 	DEVICE_DT_INST_DEFINE(id, gpio_rts5912_init_##id, NULL, &gpio_rts5912_data_##id,           \
-			      &gpio_rts5912_config_##id, POST_KERNEL, CONFIG_GPIO_INIT_PRIORITY,   \
+			      &gpio_rts5912_config_##id, PRE_KERNEL_1, CONFIG_GPIO_INIT_PRIORITY,  \
```
is this intentional? worth a mention in the commit message?

fabiobaltieri commented on drivers/gpio/gpio_rts5912.c:
```diff
@@ -412,7 +412,36 @@ static int gpio_rts5912_port_toggle_bits(const struct device *port, gpio_port_pi
+	return (gcr - (uint32_t)(RTS5912_GPIOA_REG_BASE)) / 4 + gpio->pin;
+}
+
+uint32_t gpio_rts5912_get_port_address(const struct gpio_dt_spec *gpio)
```
so `reg_base` starts as a `volatile uint32_t *` in `gpio_rts5912.c`, then get casted and returned to `uint32_t` here and then casted again as `uint32_t` in the next commit:

```c
	rts5912_rx_wake_reg = (uint32_t *)gpio_rts5912_get_port_address(&uart_rx_wakeup);
```

should it just stay `uint32_t *` all along? or `void *` I guess

JasonLin-RealTek replied:
Yes, it should be uint32_t * all along, fixed.

fabiobaltieri commented on drivers/serial/uart_realtek_rts5912.c:
```diff
@@ -31,6 +34,76 @@ struct uart_rts5912_device_config {
 
+DT_INST_FOREACH_STATUS_OKAY(PINCTRL_DT_INST_DEFINE);
+#if defined(CONFIG_PM)
+#include <reg/reg_gpio.h>
+#include "zephyr/drivers/gpio/gpio_rts5912.h"
```
can this be included all the time? we normally avoid mid-file includes unless it's like a blob struct or something like that

fabiobaltieri commented on drivers/serial/uart_realtek_rts5912.c:
```diff
@@ -31,6 +34,76 @@ struct uart_rts5912_device_config {
+#if defined(CONFIG_PM)
+#include <reg/reg_gpio.h>
+#include "zephyr/drivers/gpio/gpio_rts5912.h"
+uint32_t rx_wakeup_pin_num;
+uint32_t *rts5912_rx_wake_reg;
+struct gpio_dt_spec uart_rx_wakeup = GPIO_DT_SPEC_GET(DT_NODELABEL(uart0_wrapper), gpios);
+struct k_work_delayable rx_refresh_timeout_work;
```
`static` (all of them?)

JasonLin-RealTek replied:
We put them to the device data now.

fabiobaltieri commented on drivers/serial/uart_realtek_rts5912.c:
```diff
@@ -31,6 +34,76 @@ struct uart_rts5912_device_config {
+static void uart_rx_wait(const struct device *dev, void *user_data)
+{
+	ARG_UNUSED(user_data);
+#ifdef CONFIG_UART_CONSOLE_INPUT_EXPIRED
```
can this and the other be a `if (IS_ENABLED(CONFIG_UART_CONSOLE_INPUT_EXPIRED)) {`?

fabiobaltieri commented on boards/realtek/rts5912_evb/rts5912_evb.dts:
```diff
@@ -41,6 +41,7 @@
 	status = "okay";
 	pinctrl-0 = <&uart_rx_gpio113 &uart_tx_gpio114>;
 	pinctrl-names = "default";
+	gpios = <&gpioh 1 0>;
```
how does one go from, say `gpio113` to `gpioh1`? anyway this should probably be called `rx-gpios` or something more descriptive. The for the last zero it's common practice to use `PWM_POLARITY_NORMAL` instead.

dsseng commented on drivers/gpio/gpio_rts5912.c:
```diff
@@ -412,7 +412,36 @@ static int gpio_rts5912_port_toggle_bits(const struct device *port, gpio_port_pi
+{
+	const struct device *dev = gpio->port;
+	const struct gpio_rts5912_config *config = dev->config;
+	volatile uint32_t gcr = (uint32_t)config->reg_base;
```
Sorry, what is supposed to happen here? Why `volatile` if you copy the register value (?) into a normal stack variable? Semantically you would typically want volatile when you dereference a pointer to a hardware register or the variable might e.g. be changed by an ISR

JasonLin-RealTek replied:
Yes, the normal stack variable should not use `volatile`. Fixed

dsseng commented on dts/arm/realtek/ec/rts5912.dtsi:
```diff
@@ -31,11 +32,6 @@
 				min-residency-us = <100000000>;
 			};
 
-			suspend_to_ram: suspend_to_ram {
-				compatible = "zephyr,power-state";
-				power-state-name = "suspend-to-ram";
-				min-residency-us = <250000000>;
-			};
```
Is this power state nonfunctional? Please break out into a new commit explaining the change

JasonLin-RealTek replied:
This power state didn't used now, heavy sleep can save more power. 
We can mention it in the SOC commit.

dsseng commented on soc/realtek/ec/common/pinctrl_soc.h:
stray change

JasonLin-RealTek replied:
TYPO, we removed now, thanks.

dsseng commented on drivers/serial/uart_realtek_rts5912.c:
```diff
@@ -31,6 +34,76 @@ struct uart_rts5912_device_config {
+#include "zephyr/drivers/gpio/gpio_rts5912.h"
+uint32_t rx_wakeup_pin_num;
+uint32_t *rts5912_rx_wake_reg;
+struct gpio_dt_spec uart_rx_wakeup = GPIO_DT_SPEC_GET(DT_NODELABEL(uart0_wrapper), gpios);
```
Could you please elaborate why is this hard-coded and not sourced from DT? Thanks

keith-zephyr commented on dts/bindings/serial/realtek,rts5912-uart.yaml:
```diff
@@ -22,3 +22,6 @@ properties:
   pinctrl-names:
     required: true
+
+  gpios:
```
This should be `receive-gpios` or `rx-gpios`.  The property also needs a description field specifying how this property is used.

JasonLin-RealTek commented on soc/realtek/ec/rts5912/reg/reg_system.h:
```diff
@@ -334,4 +334,10 @@ typedef struct {
+#define RTS5912_TRNG_OFFSET 0x600UL
+#define RTS5912_OTP_OFFSET  0x680UL
+#define RTS5912_CFG_PDSTB   0x0024UL
+#define RTS5912_RNG_RUN     0x0020UL
```
Fixed

JasonLin-RealTek commented on drivers/gpio/gpio_rts5912.c:
```diff
@@ -412,7 +412,36 @@ static int gpio_rts5912_port_toggle_bits(const struct device *port, gpio_port_pi
+
+	gcr[pin_num] &= ~GPIO_GCR_MFCTRL_Msk;
+	gcr[pin_num] |= GPIO_GCR_INTSTS_Msk;
+	gcr[pin_num] |= 0x1 << GPIO_GCR_INTCTRL_Pos;
```
Fixed, thanks.

JasonLin-RealTek commented on include/zephyr/drivers/gpio/gpio_rts5912.h:
```diff
@@ -0,0 +1,39 @@
+	volatile uint32_t *gcr =
+		(volatile uint32_t *)(((uint32_t *)(DT_REG_ADDR(DT_NODELABEL(gpioa)))) + pin_num);
+
+	*gcr &= ~(GPIO_GCR_MFCTRL_Msk | GPIO_GCR_DIR_Msk);
+	*gcr |= BIT(GPIO_GCR_INTCTRL_Pos) | GPIO_GCR_INTSTS_Msk | GPIO_GCR_INTEN_Msk |
+		GPIO_GCR_INDETEN_Msk;
```
For more safety way to set the GPIO in rts5912, we will unset the multi-function bit before the GPIO interrupt setting to avoid the mistrigger of the interrupt.

JasonLin-RealTek commented on drivers/serial/uart_realtek_rts5912.c:
```diff
@@ -50,11 +159,68 @@ static int rts5912_uart_init(const struct device *dev)
+	const struct device *uart_device = dev_cfg->uart_dev;
+	const struct uart_ns16550_dev_config *uart_cfg = uart_device->config;
+	uint32_t iir = *(volatile uint32_t *)(uart_cfg->port + REG_IIR) & 0x0f;
+	uint32_t temp = 0;
```
The original intention is to read clear the register, now we use `sys_read32` to achieve this function.

JasonLin-RealTek commented on drivers/serial/uart_realtek_rts5912.c:
```diff
@@ -50,11 +159,68 @@ static int rts5912_uart_init(const struct device *dev)
+			temp = *(volatile uint32_t *)(uart_cfg->port + REG_LSR);
+			break;
+		case IIR_RBRF: /* Received data available*/
+			*(volatile uint32_t *)(uart_cfg->port + REG_IIR) |= IIR_THRE;
```
Now we use `sys_read32/ sys_write32` to do the register write/ read.
Thanks!

JasonLin-RealTek committed:
serial: uart_ns16550: add uart_ns16550_get_port() function

For user to get the register address of the device by this function

Signed-off-by: Lin Yu-Cheng <lin_yu_cheng@realtek.com>

kartben commented on drivers/serial/uart_realtek_rts5912.c:
```diff
@@ -50,11 +131,66 @@ static int rts5912_uart_init(const struct device *dev)
+			sys_read32(uart_reg + REG_LSR);
+			break;
+		case IIR_RBRF: /* Received data available*/
+			sys_write32(sys_read32(uart_reg + REG_IIR) | IIR_THRE, uart_reg + REG_IIR);
```
I doubt this comment is correct but it would be really impressive if it is :) please reopen and address if that's the case :)

keith-zephyr commented on drivers/serial/uart_realtek_rts5912.c:
```diff
@@ -50,11 +131,66 @@ static int rts5912_uart_init(const struct device *dev)
+	const uint32_t uart_reg = uart_ns16550_get_port(dev_cfg->uart_dev);
+	uint32_t iir = sys_read32(uart_reg + REG_IIR) & 0x0f;
+
+	while (iir != IIR_NIP) {
```
This loop is unbounded.  Please add a timeout check.

Can you also add a comment about why the polling of the IIR register is needed during the UART initialization.

keith-zephyr commented on dts/bindings/serial/realtek,rts5912-uart.yaml:
```diff
@@ -22,3 +22,22 @@ properties:
+      gpiog = 96 ~ 111,
+      gpioh = 112 ~ 127,
+      gpioa = 128 ~ 131
+    type: phandle-array
```
Realtek landed macros to translate the pin name to the GPIO port and pin number.
https://github.com/zephyrproject-rtos/zephyr/blob/73ea9419c354526edf175607c5f0713f892a5196/include/zephyr/dt-bindings/gpio/realtek-gpio.h#L57

Please update the help text to use the new macros. For example:
```
   rx-gpios = <RTS5912_GPIO013 0>;
```
But I recommend you select a GPIO that can be used as a UART RX pin.

JasonLin-RealTek commented on drivers/serial/uart_realtek_rts5912.c:
```diff
@@ -50,11 +131,66 @@ static int rts5912_uart_init(const struct device *dev)
+		case IIR_RBRF: /* Received data available*/
+			sys_write32(sys_read32(uart_reg + REG_IIR) | IIR_THRE, uart_reg + REG_IIR);
+			break;
+		case IIR_MASK:
```
Fixed

keith-zephyr commented on drivers/serial/uart_realtek_rts5912.c:
```diff
@@ -50,11 +134,85 @@ static int rts5912_uart_init(const struct device *dev)
+		}
+
+		iir = sys_read32(ADDR_IIR) & IIR_MASK;
+		if (iir == IIR_NIP) {
+			continue;
+		} else {
+			Z_SPIN_DELAY(10);
```
This is easier to read if you just break out of the loop directly once there is no interrupt.
```
while (_wf_cycle_count > (_wf_now - _wf_start)) {
...
    if (iir == IIR_NIP) {
        break;
    }
    k_busy_wait(10);
}
```

keith-zephyr commented on drivers/serial/uart_realtek_rts5912.c:
```diff
@@ -50,11 +134,85 @@ static int rts5912_uart_init(const struct device *dev)
+#if defined(CONFIG_PM)
+	const uint32_t uart_reg = uart_ns16550_get_port(dev_cfg->uart_dev);
+	uint32_t iir = sys_read32(ADDR_IIR) & IIR_MASK;
+	uint32_t _wf_cycle_count = k_us_to_cyc_ceil32(RTS5912_MAXIMUM_UART_POLLING_TIME_US);
```
Non-blocking comment, the leading underscores for the local variable names is not needed.

keith-zephyr commented on drivers/serial/uart_realtek_rts5912.c:
```diff
@@ -50,11 +134,84 @@ static int rts5912_uart_init(const struct device *dev)
+	 *	change GPIO pin from uart funciton to gpio function before WFI.
+	 *	System need to ensure this register is cleared in every time doing init function.
+	 */
+	while (iir != IIR_NIP && (wf_cycle_count > (wf_now - wf_start))) {
```
I recommend dropping the initialization of `iir` above and putting the only read of IIR inside the loop

```suggestion
	while (wf_cycle_count > (wf_now - wf_start)) {
		iir = sys_read32(ADDR_IIR) & IIR_MASK;
		if (iir == IIR_NIP) {
			break;
		}
		...
```

keith-zephyr commented on drivers/serial/Kconfig.realtek_rts5912:
```diff
@@ -14,7 +14,7 @@ config UART_RTS5912
 config UART_RTS5912_INIT_PRIORITY
 	int "RTS5912 UART wrapper init priority"
-	default 49
+	default 51
```
Because you've changed `UART_RTS5912_INIT_PRIORITY` here, do you still need to switch the rts5912_gpio driver init priority?

keith-zephyr commented on drivers/gpio/gpio_rts5912.c:
```diff
@@ -564,7 +582,7 @@ static DEVICE_API(gpio, gpio_rts5912_driver_api) = {
 	};                                                                                         \
 	DEVICE_DT_INST_DEFINE(id, gpio_rts5912_init_##id, NULL, &gpio_rts5912_data_##id,           \
-			      &gpio_rts5912_config_##id, POST_KERNEL, CONFIG_GPIO_INIT_PRIORITY,   \
+			      &gpio_rts5912_config_##id, PRE_KERNEL_1, CONFIG_GPIO_INIT_PRIORITY,  \
```
See the comment in the last commit.  Is this change needed because the UART driver priority got lowered?

JasonLin-RealTek replied:
Uart and espi driver will use gpio device information (example: cs_pin, uartrx_pin), therefore Zephyr compile environment need gpio driver to be initialized before uart and espi driver. 

The uart_init is at the `PRE_KERNEL_1 `area, so we move the gpio init to `PRE_KERNEL_1 `
